smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,205 @@
+    }
+
+    private fun renderInit() {
+        writer.withBlock("init {", "}") {
+            params.forEach {
+                if (it.isRequired) {
+                    write("""requireNotNull(#1L) { "endpoint provider parameter #1L is required" }""", it.name)
+                }
+            }
+        }
+    }
+
+    private fun renderEquals() {
+        writer.withBlock("public override fun equals(other: Any?): Boolean {", "}") {
+            write("if (this === other) return true")
+            write("if (other !is #L) return false", CLASS_NAME)
+            params.forEach {
+                ensureSuppressDeprecation(it)
+                write("if (this.#1L != other.#1L) return false", it.name)
+            }
+            write("return true")
+        }
+    }
+
+    private fun renderHashCode() {
+        writer.withBlock("public override fun hashCode(): Int {", "}") {
+            if (params.isEmpty()) {
+                write("return 0")
+                return@withBlock
+            }
+
+            ensureSuppressDeprecation(params[0])
+            write("var result = #L?.hashCode() ?: 0", params[0].name)
+            params.drop(1).forEach {
+                ensureSuppressDeprecation(it)
+                write("result = 31 * result + (#L?.hashCode() ?: 0)", it.name)
+            }
+            write("return result")
+        }
+    }
+
+    private fun renderToString() {
+        writer.withBlock("public override fun toString(): String = buildString {", "}") {
+            write("append(\"#L(\")", CLASS_NAME)
+            params.forEachIndexed { index, it ->
+                ensureSuppressDeprecation(it)
+                write(
+                    if (index < params.size - 1) "append(\"#1L=$#1L,\")" else "append(\"#1L=$#1L\")",
+                    it.name,
+                )
+            }
+            write("append(\")\")")
+        }
+    }
+
+    private fun renderCopy() {
+        writer.withBlock("public fun copy(block: Builder.() -> Unit = {}): #L {", "}", CLASS_NAME) {
+            withBlock("return Builder().apply {", "}") {
+                params.forEach {
+                    ensureSuppressDeprecation(it)
+                    write("#1L = this@#2L.#1L", it.name, CLASS_NAME)
+                }
+            }
+            write(".apply(block).build()")
+        }
+    }
+
+    private fun renderBuilder() {
+        writer.withBlock("public class Builder internal constructor() {", "}") {
+            params.forEach {
+                it.renderDeclaration(writer, it.defaultLiteral, isMutable = true)
+            }
+            write("")
+            write("internal fun build(): #1L = #1L(this)", CLASS_NAME)
+        }
+    }
+}
+
+// kotlin-mapped representation of an endpoint parameter
+private data class KotlinEndpointParameter(
+    val name: String,
+    val type: Symbol,
+    val isRequired: Boolean,
+    // All endpoint params are nullable. Standard symbol rendering suppresses the default value for boxed types, but
```

Comment:
Sort of, this could be changed. You can also register a new property formatter on the writer if warranted. What you have is fine though, just clarifying this isn't set in stone or how it must work or anything